[2단계 - 자동차 경주] 이태순 미션 제출합니다.

LTSGOD opened the pull request.

LTSGOD committed:
refactor: 1단계 피드백 반영

- 불필요한 static 제거
- 코딩 컨벤션 수정

LTSGOD committed:
refactor: model, domain, controller 패키지 생성

- 파일 이동

LTSGOD committed:
chore: 클래스 이름 변경

LTSGOD committed:
refactor: 게임 생성 테스트 및 리펙토링 완료

LTSGOD committed:
feat: 게임 종료 메서드 추가

- 게임 종료 isEnd()메서드 추가 및 테스트 완료

LTSGOD committed:
refactor: domain 과 view 로직 분리

LTSGOD committed:
chore: 코딩 컨벤션 수정

LTSGOD committed:
chore: Output -> OutputView 로 클래스 이름 변경

- printRound -> printCars 로 메서드 이름 변경

LTSGOD committed:
refactor(RandomNum): 랜덤 생성 기능 Car 클래스와분리

- RandomNum class 생성 테스트 완료

LTSGOD committed:
refactor(CarTest): Random 생성 테스트 분리

LTSGOD committed:
refactor(Car): Random 수 생성 분리에 따른 메서드 수정

- move 메서드에 파라미터로 넘겨주는 것으로 수정

LTSGOD committed:
refactor: Random 수 생성 기능 Car와 분리 및 테스트 완료

LTSGOD committed:
fix(InputView): Input 버그 수정

LTSGOD committed:
feat(CarTest): 불필요한 테스트 제거

LTSGOD committed:
chore: 코드 컨벤션 수정

LTSGOD committed:
feat: move method private 전환

LTSGOD committed:
feat: GameTest 테스트코드 수정

LTSGOD committed:
chore: 기존 테스트 클래스 위치 변경

LTSGOD committed:
feat: 예외처리 테스트 추가

LTSGOD committed:
feat: CarTest race메서드 테스트

LTSGOD committed:
feat: 중복 이름 예외 처리 추가

- 중복 이름 예외처리에 따라 메서드 이름 변경
  (isLong()->checkValidation())
- 메서드 분리 isLessThanFiveLetters, isDuplicatedName

LTSGOD committed:
feat: 이름사이에 delimiter 예외처리 추가

LTSGOD committed:
chore: 코딩 컨벤션 수정

LTSGOD committed:
chore: isEnd 변수 이름 변경 (countRound -> countEnd)

LTSGOD committed:
refactor: Input할때 while문 제거

- while 문제거 try catch 에서 호출 하는 방식으로 변경

LTSGOD committed:
refactor: getter 사용 지양

- RacingGame.java -> isSamePosition, isSameWinNum 메서드 사용
- Car.java -> isSamePosition, isSameWinNum 메서드 생성

LTSGOD committed:
chore: 코딩 컨벤션 수정

LTSGOD committed:
refactor: 코드 중복 제거

LTSGOD committed:
refactor: filter안 변수 이름 명확하게 변경

dongho108 commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -32,28 +25,29 @@ public int getWinNum() {
-	public boolean move() {
-		if (createRandomNum() >= MOVE) {
+	private void move(int randomNum) {
+		if (randomNum >= MOVE) {
```
car의 move가 random number 라는 것을 알 필요가 있을까요?
```suggestion
	private void move(int number) {
		if (number >= MOVE) {
```

dongho108 replied:
혹시 이 부분은 반영하신건가요?

dongho108 commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -32,28 +25,29 @@ public int getWinNum() {
-		return false;
+	}
+
+	public boolean isSameWinNum(int maxWin) {
+		return this.winNum == maxWin;
```
Car가 maxWin 이라는 이름을 알아야할까요? 또한, Car가 "win" 이라는 개념을 알까요? win은 RacingGame에만 있는 개념 아닐까요? 태순님의 생각은 어떠신가요?

LTSGOD replied:
요구사항을 잘못 이해해서 다른 로직을 짠것같습니다! 수정하겠습니다

dongho108 commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -32,28 +25,29 @@ public int getWinNum() {
-	public boolean move() {
-		if (createRandomNum() >= MOVE) {
+	private void move(int randomNum) {
+		if (randomNum >= MOVE) {
```
MOVE 라는 이름으로 의도를 충분히 드러낼 수 있을까요?

LTSGOD replied:
더 명확히 의도가 드러나게 수정하겠습니다

dongho108 commented on src/test/java/racingcar/domain/CarTest.java:
```diff
@@ -0,0 +1,37 @@
+	}
+
+	@Test
+	@DisplayName("정상적으로 race 테스트")
```
정상적으로 라는 말보다는 좀 더 풀어서 쓸 수 없을까요?

LTSGOD replied:
요구사항 변경후 테스트코드 수정후다시 리뷰요청하겠습니다!!

dongho108 commented on src/test/java/racingcar/util/RandomNumTest.java:
```diff
@@ -0,0 +1,22 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class RandomNumTest {
+
+	@Test
+	@DisplayName("랜덤 수 생성")
+	void createRandomNum() {
+
+		int num = RandomNum.createRandomNum();
+
+		assertAll(
+			() -> assertThat(num).isGreaterThanOrEqualTo(0),
+			() -> assertThat(num).isLessThanOrEqualTo(9)
+		);
+	}
+}
```
랜덤 수가 제대로 나오는지 테스트할 필요가 있을까요?

dongho108 commented on src/main/java/racingcar/domain/RacingGame.java:
```diff
@@ -0,0 +1,76 @@
+
+public class RacingGame {
+	private List<Car> cars;
+	private final int TRY_NUM;
```
인스턴스 변수는 소문자 써주세요!

LTSGOD replied:
넵 변경하고 다음부터 컨벤션 지키겠습니다!

dongho108 commented on src/main/java/racingcar/domain/RacingGame.java:
```diff
@@ -0,0 +1,76 @@
+		this.TRY_NUM = TRY_NUM;
+	}
+
+	public boolean isEnd() {
+		if (countEnd == TRY_NUM) {
+			return true;
+		}
+		countEnd++;
+		return false;
+	}
```
isEnd()는 끝인지 체크하는 역할도 하지만 countEnd를 늘리는 역할까지 두가지 역할을 하고있네요, countEnd를 늘리는 역할을 적절한 곳에 분리해볼 수 있을까요?

dongho108 commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -32,28 +25,29 @@ public int getWinNum() {
 
 	public void race(int num) {
 		int positionStandard = position;
 		for (int i = 0; i < num; i++) {
-			move();
+			move(RandomNum.createRandomNum());
 		}
 		position -= positionStandard;
-		Output.printRace(name, position);
 	}
```
어.. 다시보니까 요구사항에 맞게 개발이 안된것 같습니다..! 매 회마다 4회씩 전진하는게 아니라 한번에 하나씩 전진하면서 총 4번을 전진시도를 하는건데요! 요구사항 다시 읽어보시고 맞게 수정해서 다시 제출해주셔야 할 것 같아요! 이해 안되시면 DM 주셔도 됩니다.

LTSGOD replied:
아 넵 요구사항을 잘못 이해했네요! 다시 리펙토링 해서 리뷰요청 하겠습니다.

dongho108 reviewed (changes requested):
안녕하세요 태순님!! 2단계 구현하시느라 고생하셨습니다.

코드가 점점 발전되는 모습을 보는 것 같아서 좋았습니다.
그런데 중간에 요구사항에 안맞게 구현하신 부분을 찾아서 이 부분을 제대로 다시 구현하신 후에 제출하신 코드를 리뷰하는게 맞다고 생각이 들더라구요! 그래서 다시한번 수정하고 요청주시면 빠르게 리뷰 해드리겠습니다!

LTSGOD committed:
refactor(Car, RacingGame): 요구사항에 맞게 다시 리펙터

- WinNum 관련 메서드,변수 제거
- RacingGame 에서 한라운드에서 tryNum 횟수 만큼 전진 ->
  한라운드에서1번씩 전진

LTSGOD committed:
refactor: 상수이름 더 명확히 변경

LTSGOD committed:
refactor: 인스턴수 변수 소문자로 변경

LTSGOD committed:
refactor: tryNum 로직 변경

- tryNum을 굳이 넘겨주지 않게 하여 불필요한 메서드, 변수 제거
- isEnd() 메서드 제거, countEnd 인스턴스 변수 제거
- 인스턴스 변수 car final지정

LTSGOD committed:
refactor: 필요없는 test 제거

LTSGOD committed:
refactor: test 용이하게 Random Interface 생성

- Car의 move 메서드 수정

LTSGOD committed:
refactor: Random 메서드 구조 변경에 따른 코드 변경

LTSGOD committed:
refactor(Car): move 메서드 변경

- Random 주입 받게 변경

LTSGOD committed:
refactor: 요구사항에 맞춰 RacingGameTest변경

LTSGOD committed:
chore: 상수에 static 추가, 필요없는 라이브러리 제거

LTSGOD committed:
refactor: 요구사항 수정

dongho108 commented on src/main/java/racingcar/util/Random.java:
```diff
@@ -0,0 +1,5 @@
+package racingcar.util;
+
+public interface Random {
+	int generator();
```
메서드 명이 동사로 시작하지 않고 있어요!

dongho108 commented on src/main/java/racingcar/util/Random.java:
```diff
@@ -0,0 +1,5 @@
+package racingcar.util;
+
+public interface Random {
```
인터페이스의 역할과 이름이 잘 지어져 있나요?

LTSGOD replied:
역할에 충실한 이름을 짓지 못한 것 같네요 좀 더 고민해보고 수정해보겠습니다!

dongho108 commented on src/main/java/racingcar/domain/RacingGame.java:
```diff
@@ -0,0 +1,50 @@
+		return selectFinalWinner(rankFinal(cars));
+	}
+
+	public List<Car> selectFinalWinner(List<Car> carRanking) {
+		int maxPosition = carRanking.get(0).getPosition();
```
만약 비어있는 리스트가 파라미터로 전달된다면 어떤 문제가 생길 수 있을까요?

LTSGOD replied:
의도했던대로 로직이 안돌아가서 프로그램이 망가질 수 있다고 생각했는데 이게 맞을까요?? 맞다면 비어있는 리스트를 다루는 코드를 추가 해야하는게 맞나요??

dongho108 replied:
비어있는 리스트에서 `.get(0)`으로 첫번째 원소에 접근하려고 하면 예외가 발생할 위험이 있어요.

`비어있는 리스트를 다루는 코드를 추가 해야하는게 맞나요??` 처럼 비어있는 리스트인지 검증하는 코드가 있어야할 필요도 있겠죠~

dongho108 commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,32 @@
+		return this.position == position;
+	}
+
+	public void move(Random RandomNum) {
+		if (RandomNum.generator() >= CRITICAL_POINT_OF_MOVING) {
```
RandomNum 이라는 변수명이 의미를 잘 표현하고 있을까요? 개인적으로 ```RandomNum.generator()``` 라는 코드가 어색하게 느껴지는데 태순님은 어떻게 생각하시나요?

LTSGOD replied:
좀 더 자연스러운 네이밍을 해보도록 노력하겠습니다!

dongho108 replied:
이 부분 반영 되었을까요?

dongho108 commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -1,54 +1,50 @@
 	}
 
-	private static boolean noExceptionAtNum(String input) {
+	private static void intInputProcess(Scanner sc) {
```
메서드명이 동사로 시작하지 않고 있어요!

dongho108 commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -1,54 +1,50 @@
 	}
 
-	private static boolean isExceptionAtStrings(String input) {
+	private static void stringInputProcess(Scanner sc) {
```
메서드명이 동사로 시작하지 않고 있어요!

LTSGOD replied:
동사로 시작하도록 메서드이름을 수정하겠습니다

dongho108 replied:
이 부분 반영 되었을까요?

dongho108 commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -1,54 +1,50 @@
 	}
 
-	private static boolean noExceptionAtNum(String input) {
+	private static void intInputProcess(Scanner sc) {
```
변수명은 축약하지 않을 수 있을까요?

LTSGOD replied:
이름이 쓸데없이 길었던것 같네요 축약해보겠습니다

dongho108 replied:
`sc` 같이 축약을 하지 말아야 한다고 말씀드린거에요 ㅎㅎ!

dongho108 commented on src/main/java/racingcar/view/RacingCarInputException.java:
```diff
@@ -0,0 +1,56 @@
+		}
+	}
+
+	private static void isDelimiterInName(String s) {
+		if (s.contains(",")) {
+			throw new IllegalArgumentException("이름 사이에는 ,가 들어갈 수 없습니다.");
+		}
+	}
+
+	private static void isLessThanFiveLetters(String s) {
+		if (s.length() > MAX_LENGTH) {
+			throw new IllegalArgumentException("5이하의 문자열을 입력하세요.");
+		}
+	}
+
+	private static void isDuplicatedName(ArrayList<String> strings, String s) {
+		if (strings.contains(s)) {
+			throw new IllegalArgumentException("중복된 이름입니다. 다른이름을 입력하세요.");
+		}
+	}
+
+	public static void isPositive(int num) {
+		if (num < 0) {
+			throw new IllegalArgumentException("1회 이상의 수를 입력하세요.");
+		}
+	}
+
+	public static int isNum(String input) {
+		return Integer.parseInt(input);
+	}
```
이 메서드들이 ```is~``` 로 시작하는 네이밍이 맞을까요? 내부 동작을 보면 true false를 판단하는게 아니라 예외를 뱉고있어요. 태순님 생각은 어떠신가요?

LTSGOD replied:
예외처리를 다루니 handle~로 시작하는게 더 바람직 할것같다는 생각이 드네요!

dongho108 commented on src/test/java/racingcar/domain/CarTest.java:
```diff
@@ -0,0 +1,52 @@
+	}
+
+	@Test
+	@DisplayName("4미만의 숫자에 car의 안움직이는지 테스트")
+	void stop() {
```
저는 이 문장이 어색하다고 느껴지는데 더 가독성 있게 바꿔볼 수 있을까요?

LTSGOD replied:
수정해보겠습니다!

dongho108 commented on src/test/java/racingcar/domain/RacingGameTest.java:
```diff
@@ -0,0 +1,70 @@
+class RacingGameTest {
+
+	RacingGame game;
+	final ArrayList<String> NAMES = new ArrayList<>(Arrays.asList("na", "tae", "joe"));
+
```
변수명을 대문자로 지으신 이유가 있을까요?

LTSGOD replied:
코드를 짜면서 잠시 혼동한거 같습니다! 다시 수정할게요!

LTSGOD replied:
코드를 짜면서 잠시 혼동한거 같습니다! 다시 수정할게요!

dongho108 commented on src/test/java/racingcar/domain/RacingGameTest.java:
```diff
@@ -0,0 +1,70 @@
+	}
+
+	@Test
+	@DisplayName("최종 우승자가 여러명이면 공동우승자를 뽑는다")
+	void selectFinalWinner() {
```
문장형으로 이름을 짓는 테스트와 아닌 테스트들이 혼재되어있는데 하나도 통일할 수 있을까요?

LTSGOD replied:
아넵 한개의 형식으로 통일하겠습니다.

dongho108 commented on src/test/java/racingcar/domain/RacingGameTest.java:
```diff
@@ -0,0 +1,70 @@
+	void selectFinalWinner() {
+		//given
+		List<Car> cars = game.createCars(new ArrayList<>(Arrays.asList("na", "tae", "joe")));
+		for (int i = 0; i < 3; i++) {
+			game.race();
+		}
```
테스트 메서드 안에서는 반복문을 쓰기보단 픽스처로 분리하여 가독성있게 리팩토링 하는것을 추천드립니다~

LTSGOD replied:
아 넵 픽스처로 분리하겠습니다!

dongho108 commented on src/test/java/racingcar/domain/RacingGameTest.java:
```diff
@@ -0,0 +1,70 @@
+		//when
+		List<Car> winner = game.selectFinalWinner(game.rankFinal(cars));
+
+		assertThat(winner.size()).isEqualTo(
+			cars.stream().filter(m -> m.getPosition() == winner.get(0).getPosition()).count());
```
검증 코드안에서는 로직을 최대한 빼주세요~ 어떤 테스트를 하고있는지 한번에 알기 힘들어요! 변수로 빼고 이름을 지어주시는게 어떨까요?

LTSGOD replied:
아 넵 분리해서 더 가독성 좋게 수정하겠습니다

dongho108 commented on src/test/java/racingcar/view/RacingCarInputExceptionTest.java:
```diff
@@ -0,0 +1,69 @@
+	void isNull() {
+		String test = null;
+
+		Assertions.assertThrows(IllegalArgumentException.class, () -> RacingCarInputException.isNull(test));
+	}
```
static import를 해주시면 더 깔끔할 것 같습니다.
```suggestion
		assertThrows(IllegalArgumentException.class, () -> RacingCarInputException.isNull(test));

```

LTSGOD replied:
넵 수정하겠습니다!

dongho108 reviewed (changes requested):
태순님 리뷰 반영하느라 고생하셨습니다!
몇가지 코멘트 드렸으니 리뷰 반영하고 다시 리뷰요청 해주세요! 😄

LTSGOD committed:
refactor: 인터페이스 역할에 따라 메서드명 renaming

LTSGOD committed:
refactor: 인터페이스, 구현체 이름 변경

- NumGenerator, ZeroToNineNumGenerator

LTSGOD committed:
refactor: 인터페이스 이름 변경 및 메서드 이름 변경

LTSGOD committed:
refactor: 예외처리 메서드 이름 변경

LTSGOD committed:
refactor: test형식 통일, 변수 이름 대문자-> 소문자 변경, 중복 부분 픽스처로 분리

LTSGOD committed:
refactor: 인터페이스, 구현체 분리에따른 이름 변경

dongho108 commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -1,54 +1,50 @@
 	}
 
-	private static boolean isExceptionAtStrings(String input) {
+	private static void stringInputProcess(Scanner sc) {
```
메서드명이 동사로 시작하지 않고있어요~

dongho108 commented on src/main/java/racingcar/util/ZeroToNineNumGenerator.java:
```diff
@@ -0,0 +1,9 @@
+package racingcar.util;
+
+public class ZeroToNineNumGenerator implements NumGenerator {
+    private static final int ZERO_TO_NINE_CONVERTER = 10;
```
만약 랜덤 수가 11이 된다면 클래스 이름까지 변경이 되어 이 클래스를 사용하는 모든 클래스에서 변경이 일어날 수 있을 것 같아요!
자료구조를 변수명에 넣지 않는 것처럼 클래스명도 바뀔 수 있는 부분에 대한 이름을 넣지 않는것이 좋습니다~

LTSGOD replied:
넵 이름을 수정하겠습니다

dongho108 commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -0,0 +1,50 @@
+        }
+    }
+
+    private static void validateNames(Scanner sc) {
+        try {
```
변수명은 축약하지 말아주세요!

dongho108 commented on src/main/java/racingcar/domain/RacingGame.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    public List<Car> selectFinalWinner(List<Car> carRanking) {
+        int maxPosition = carRanking.get(0).getPosition();
+
```
리스트의 요소를 인덱스로 직접 접근하는 코드는 예외를 뱉을 확률이 높은 코드여서 이와 같은 상황은 검증 코드가 필요할 수도 있어요~

dongho108 commented on src/test/java/racingcar/domain/CarTest.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    @Test
+    @DisplayName("4미만의 숫자에 car의 안움직이는지 테스트")
```
이 부분 반영이 되었을까요?

dongho108 commented on src/test/java/racingcar/domain/RacingGameTest.java:
```diff
@@ -0,0 +1,69 @@
+        List<Car> cars = game.getCars();
+        //when
+        List<Car> winner = game.selectFinalWinner(game.rankFinal(cars));
+        long winnerNum = cars.stream().filter(m -> m.getPosition() == winner.get(0).getPosition()).count();
```
한줄에는 하나의 점만 찍어주세요!

dongho108 reviewed (approved):
반영을 하지 않으신 부분도 있는데 특별한 이유가 있을까요~?

미션 기간이 겹치게 되면 다음 미션때 힘들어질 수 있기 떄문에 일단 머지하겠습니다!
몇가지 코멘트를 더 남겼는데 꼭 확인해주시면 감사하겠습니다.


길고 긴 자동차 미션 하느라 고생 정말 많으셨습니다!
다음 미션에서 뵙겠습니다~~!! 👏
